### DISCUSSION TITLE
How to use mlflow with ultralytics ?

### DISCUSSION THREAD

**Discussion by Ashingharoy1991** [ultralytics/ultralytics#17490] [Docs]
I've been exploring the Ultralytics documentation and saw that MLflow integration is possible by adjusting some settings. However, I’m encountering an issue when running training scripts in Python.

When I specify project and name arguments during training, the mlruns directory doesn't get created. But, if I skip these parameters, MLflow automatically creates the mlruns directory in the runs folder. Why does this happen? Is there a way to get consistent behavior with project and name specified?

Additionally, I’d like to change the default path for MLflow’s mlruns directory to a custom location. Is there a way to configure this through Ultralytics directly? Or I have to use mlflow?

Any insights would be appreciated!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Ashingharoy1991, thank you for reaching out with your questions about integrating MLflow with Ultralytics 🚀! We're here to help guide you in the right direction.

We recommend checking out our [Docs](https://docs.ultralytics.com) for an overview of features and configurations. Specifically, for logging and integration details, you might find the [Python usage examples](https://docs.ultralytics.com/usage/python/) helpful.

Since you're encountering an issue that could be a bug, it would be great if you could provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) of your script, including the configuration that leads to the behavior you’ve described. This will help us assist you more effectively.

For custom training setups and configuration questions, providing more details such as dataset specifics, your training commands, and any relevant logs can be incredibly useful. Make sure you're also following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Feel free to join our community in real-time on [Discord](https://ultralytics.com/discord) 🎧 or share your experiences and get insights on our [Discourse](https://community.ultralytics.com). Our [Subreddit](https://reddit.com/r/ultralytics) is also a great place for deeper dives into community knowledge sharing.

## Upgrade

Make sure you're using the latest version of the `ultralytics` package to ensure all enhancements and bug fixes are included. Install or upgrade with:

```bash
pip install -U ultralytics
```

## Supported Environments

YOLO works in several verified environments. Check out:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: Explore our [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon AWS**: Review the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: See our [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

The status badge indicates if all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing, confirming the integrity of various [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across different operating systems.

This is an automated response, and an Ultralytics engineer will join the conversation soon to provide further assistance. 🙌

**Reply by Ashingharoy1991** [reply]
https://docs.ultralytics.com/integrations/mlflow/#introduction
 In this section it's mentioned,how can I use mlflow

**Reply by glenn-jocher** [reply]
@Ashingharoy1991 to use MLflow with Ultralytics YOLO, ensure MLflow is installed, enable it in the settings, and start a local MLflow server for tracking.

**Reply by Ashingharoy1991** [reply]
@glenn-jocher  hello!
I'm facing a problem with organizing the output from my YOLO training pipeline. I’ve set the project name and name parameters as suggested, expecting YOLO to create a structured directory with an mlrun folder and organize the training results under the specified project and run names. However, the following issues occur:

1. After training, the mlrun directory is not created within the specified project folder.


2. The results are not being reflected on the server as expected

**Reply by glenn-jocher** [reply]
Hello! It seems like there might be a configuration issue. Please ensure that MLflow logging is enabled in your settings and that the `MLFLOW_TRACKING_URI` is correctly set to point to your server. If the problem persists, try updating your Ultralytics package to the latest version.

**Comment by Ashingharoy1991** [comment]
![image](https://github.com/user-attachments/assets/adba49be-0a6f-4ae4-bcbd-50eb577d0370)
![image](https://github.com/user-attachments/assets/2ea67d12-b24e-4b9b-9604-6c31c29c2fe6)
![image](https://github.com/user-attachments/assets/b1f480fa-3a70-4fd7-bd21-6f7050d100ed)
 @glenn-jocher  hi, actually this is the step i am applying to used mlflow.. actually while i am doing training without specified project and name arguments , on that time in run folder the mlrun directory created and it store all information about run...
 
 
 also i am using this code. but don't know how my pc get stuck in between the training
 
 from ultralytics import YOLO
import mlflow
import os

# Define the working directory for storing model artifacts
working_dir = "/home/st1/Mlops"  # Define your working directory here

# # Define the experiment ID for MLflow (make sure 'exp' is defined or replace it with an integer)
# exp = 0  # Replace this with your actual experiment ID

# Start an MLflow run
with mlflow.start_run(experiment_id=501178566973958999):
    # Log data source information
    mlflow.log_param("data_source", "MinIO")
    
    # Define the YOLO model and start training
    model = YOLO("yolov8n.pt")
    
    # Set the path to the `data.yaml` file
    data_yaml_path = "/home/st1/Mlops/yolo_v8.yaml"  # Replace with actual path if not dynamically generated
    results = model.train(data=data_yaml_path, epochs=2, batch=16)
    
    # Retrieve the metrics dictionary from the model
    metrics = model.metrics.results_dict  # Access the dictionary of metrics
    
    # Print the metrics dictionary for debugging
    print("Metrics:", metrics)
    
    # Log the relevant metrics to MLflow if they exist
    precision = metrics.get('metrics/precision(B)', None)
    if precision is not None:
        mlflow.log_metric("precision", precision)

    recall = metrics.get('metrics/recall(B)', None)
    if recall is not None:
        mlflow.log_metric("recall", recall)

    mAP50 = metrics.get('metrics/mAP50(B)', None)
    if mAP50 is not None:
        mlflow.log_metric("mAP50", mAP50)

    mAP50_95 = metrics.get('metrics/mAP50-95(B)', None)
    if mAP50_95 is not None:
        mlflow.log_metric("mAP50-95", mAP50_95)

    fitness = metrics.get('fitness', None)
    if fitness is not None:
        mlflow.log_metric("fitness", fitness)

    # Log train and validation loss if available
    train_loss = metrics.get("train_loss", None)
    if train_loss is not None:
        mlflow.log_metric("train_loss", train_loss)

    val_loss = metrics.get("val_loss", None)
    if val_loss is not None:
        mlflow.log_metric("val_loss", val_loss)

    # Save and log model weights
    model_path = os.path.join(working_dir, "yolov8_model_weights.pt")  # Use working_dir for storing model
    model.save(model_path)
    
    # Log the saved model as an artifact in MLflow
    mlflow.log_artifact(model_path, artifact_path="models")

**Reply by glenn-jocher** [reply]
@Ashingharoy1991 to ensure consistent behavior with MLflow when specifying project and name, make sure these parameters are correctly set in your environment or script. To change the default path for MLflow's `mlruns` directory, set the `MLFLOW_TRACKING_URI` environment variable to your desired location before starting the MLflow server. For more details, refer to the MLflow documentation.

**Comment by Ashingharoy1991** [comment]
![image](https://github.com/user-attachments/assets/667213cd-99a6-4a62-8ca1-ef0b79b0f1d8)
@glenn-jocher  hello!

here you can see i am changing the runs_dir patha but its not changing , can help me with it?

**Reply by glenn-jocher** [reply]
@Ashingharoy1991 hello! To change the MLflow `mlruns` directory path, set the `MLFLOW_TRACKING_URI` environment variable to your desired location before starting the MLflow server.

**Reply by Ashingharoy1991** [reply]
@glenn-jocher hello thank you for your assistance.. before starting training I need to set normally the mlflow track_uri. In this case also the same? Because I have tried this but it's still created on the run directory

**Reply by glenn-jocher** [reply]
@Ashingharoy1991 hello! Yes, you should set the `MLFLOW_TRACKING_URI` environment variable before starting your training to specify a custom directory for MLflow logs. If it's still creating logs in the default directory, please ensure the variable is correctly set in your environment.

**Reply by Ashingharoy1991** [reply]
@glenn-jocher  hello! the previous issue solved, thank you for your help,
another thing i want to know.
1. while i am doing training with this code 
![image](https://github.com/user-attachments/assets/31a52eff-699f-4bb8-9b5b-02c55dfd4045)
the mlflow directory is created where i want to save the directory. now i have realised that the project arguments in yolo training actually the experiment name for mlflow and based on this project runs are store in that  experiments.. what i am facing problem that lets say set directory is "/home/sti/test_mlflo" in this path the mlflow directory created and i want to save the project name folder also here so i run the in this directory so in this path now i have two directory one is "mlflow_test" and one is "mlflow". in the "mlflow" directory all the metadata are store for mlflow and in "mlflow_test" directory i have the yolo train weight folder and corresponding  matrix images.. now if i open the mlflow server with this command "mlflow server --backend-store-uri /home/st1/test_mlflow/mlflow" then i can see all runs for the corresponding project and artifacts also , now the problem is that while i want to do model registry with the server UI , i am not getting any  registry options , also i have tryout this command 
"mlflow server --backend-store-uri /home/st1/Mlops/test_project/ --default-artifact-root /home/st1/Mlops/test_project/838328226548121284/0c976120808246a48922057e5f5b33fc/artifacts"
![image](https://github.com/user-attachments/assets/2b4cfb77-f8d6-4198-a6df-da8b7de572dc)
in this screen shots you can see there is no registry options i have
![image](https://github.com/user-attachments/assets/e447e81e-46ed-4108-9ca5-e1625f5b87df)

you can ignore the experiments name , problem is same

**Reply by glenn-jocher** [reply]
To enable model registry in MLflow, you'll need to use a proper backend database (like SQLite, PostgreSQL, or MySQL) instead of just a file system URI. Try starting the MLflow server with SQLite like this:

```bash
mlflow server --backend-store-uri sqlite:///mlflow.db --default-artifact-root /home/st1/test_mlflow/mlflow
```

This will create a SQLite database file and enable the model registry feature in the MLflow UI. The registry tab should then appear in the navigation menu.

**Reply by Ashingharoy1991** [reply]
@glenn-jocher  hello.. thank you for your reply. actually i have already try this code. server is running but my previous experiments which is store in that path that's not showing .. its only showing while i am using local path to starting the server

**Reply by glenn-jocher** [reply]
To see your previous experiments in MLflow, make sure the `--backend-store-uri` points to the exact directory containing your MLflow tracking database (mlruns). The directory structure should be maintained between runs. You can verify the correct path by checking where your `mlruns` directory and metadata files are located.

**Reply by Ashingharoy1991** [reply]
@glenn-jocher  hello, 
![image](https://github.com/user-attachments/assets/34c466bf-cb35-4ab2-b48b-c2b3e03b953d)
you can see i have set the MLFLOW_TRACKING_URI to the mlflow server url
after that i have run this code 
![image](https://github.com/user-attachments/assets/3255350f-4f62-4cb1-bdf8-c67d606b4bb8)
just to see wither its getting log or not.
i have run the mlflow server with this code 
![image](https://github.com/user-attachments/assets/a0b74e05-8ee4-4eb3-a1d7-5c872d078d44)

now all model metrics logged and i can see the artifacts also but not getting register model option
![image](https://github.com/user-attachments/assets/db324788-9109-4321-ba9c-65310bf7fe15)
 do i need to log the model? if yes then how can i add the model with ultralytics cli
 
  https://www.mlflow.org/docs/latest/model-registry.html#adding-an-mlflow-model-to-the-model-registry
  
  this is the documents i am seeing 
  also i have seen that model is not log, thats why its not giving me the register model options, is there any options to active that? or i need to used mlflow autologin functions? if yes then how can i used it with ultralytics CLI training command

**Reply by glenn-jocher** [reply]
To register a model in MLflow, you'll need to log the model as an artifact during training. Unfortunately, the Ultralytics CLI doesn't support direct MLflow model logging, so you'd need to use MLflow's `log_model` function in a custom Python script after training. For CLI-based workflows, consider using MLflow's `autolog` feature in a Python script instead. For more details, refer to the [MLflow documentation](https://www.mlflow.org/docs/latest/model-registry.html#adding-an-mlflow-model-to-the-model-registry).

**Reply by Ashingharoy1991** [reply]
@glenn-jocher Hi thank you for your reply.. just saw that today.. what if after training if I used a script with the mlflow client. Based on the experiment I will find  the run which I want to log.. and then I register the model?

**Reply by glenn-jocher** [reply]
You can use the MLflow client in a script after training to find the desired run and register your model. This approach allows you to log and manage your models efficiently.

**Reply by Ashingharoy1991** [reply]
@glenn-jocher  hello thank you for your help :)
this is the code i have written and its working but only for the model metrics i am not getting like this 
this one generated by ultralytics
![image](https://github.com/user-attachments/assets/4cc88a41-84a3-4538-9450-cf7da656255d)
this one by mine 
![image](https://github.com/user-attachments/assets/abaa8db1-8508-4453-89dc-1305917dd8c8)
![image](https://github.com/user-attachments/assets/39f6ccb0-e9d4-4c43-9222-e1d6287caa28)
![Uploading image.png…]()

**Reply by glenn-jocher** [reply]
Hello! It seems you're looking to match the format of the model metrics logged by Ultralytics. To achieve this, ensure you're logging all relevant metrics and artifacts, and consider using MLflow's `log_metrics` and `log_artifacts` functions for consistency. If you're still encountering issues, reviewing the Ultralytics code for specific logging details might help.

**Reply by Ashingharoy1991** [reply]
@glenn-jocher hello good morning!
Yes you are right , i want to achieve the format of the model matrix.. where can I find the ultralytics code ?? I searched but did not find it and also I have seen that in my code after the model train was done at that time if I print the model matrix that matrix is actually going to log.

**Reply by glenn-jocher** [reply]
Good morning! You can explore the Ultralytics codebase on our [GitHub repository](https://github.com/ultralytics/ultralytics). Look into the sections related to model training and logging to understand how metrics are being captured and logged. If you need more specific guidance, feel free to ask!
